File: studio/analysis-core/src/look.rs
//! The LOOK contract in Rust — a 1:1 mirror of `SixFour.Spec.Look` (MATH.md §9).
//! The verifiable part of the per-user look: a bounded `LookCode`, a reference
//! `affine_look` transform, and the global collapse. A future MLX-trained warp
//! is another inhabitant of the same contract (neutral=identity, gamut-closed,
//! bounded, continuous) — these properties are tested here, weight-independent.

use crate::color::{dist_sq, Oklab};
use crate::cyclic::CyclicStack;

pub const LOOK_DIM: usize = 4; // warmth, contrast, saturation, lift
pub const LOOK_BOUND_MAX: f64 = 0.8;
pub const LOOK_LIPSCHITZ: f64 = 1.5;

/// A bounded control vector in [-1,1]^LOOK_DIM. Control-surface-agnostic: named
/// knobs / latent / a transformer-on-top all map into this.
#[derive(Clone, Copy, Debug)]
pub struct LookCode(pub [f64; LOOK_DIM]);

impl LookCode {
    pub fn neutral() -> Self {
        LookCode([0.0; LOOK_DIM])
    }
    /// Rejects out-of-box values.
    pub fn new(v: [f64; LOOK_DIM]) -> Option<Self> {
        if v.iter().all(|&x| (-1.0..=1.0).contains(&x)) {
            Some(LookCode(v))
        } else {
            None
        }
    }
}

/// Reference (non-learned) look transform: a residual, clamped colour-space
/// warp. Neutral code ⇒ identity. Mirrors Haskell `affineLook`.
pub fn affine_look(code: &LookCode, palette: &[Oklab]) -> Vec<Oklab> {
    let v = code.0;
    let warmth = v[0].tanh() * 0.08;
    let contrast = 1.0 + 0.5 * v[1].tanh();
    let sat = 1.0 + 0.5 * v[2].tanh();
    let lift = v[3].tanh() * 0.10;
    let cl = |lo: f64, hi: f64, x: f64| x.max(lo).min(hi);
    palette
        .iter()
        .map(|c| {
            [
                cl(0.0, 1.0, 0.5 + (c[0] - 0.5) * contrast + lift),
                cl(-0.4, 0.4, c[1] * sat + warmth),
                cl(-0.4, 0.4, c[2] * sat + warmth),
            ]
        })
        .collect()
}

/// Collapse a per-frame stack to one global palette `G` (population-weighted
/// k-means) + per-frame local→global index remaps (nearest in OKLab). The
/// global analog of the per-frame `CompleteVoxelVolume`.
pub fn global_collapse(stack: &CyclicStack, k_global: usize, iters: usize) -> (Vec<Oklab>, Vec<Vec<u8>>) {
    let cands = crate::collapse::candidates(stack);
    let g = crate::collapse::weighted_kmeans(&cands, k_global, iters, 1);
    let remap = stack
        .frames
        .iter()
        .map(|f| {
            f.palette
                .iter()
                .map(|&c| {
                    let mut best = 0usize;
                    let mut bd = f64::INFINITY;
                    for (j, &gj) in g.iter().enumerate() {
                        let d = dist_sq(c, gj);
                        if d < bd {
                            bd = d;
                            best = j;
                        }
                    }
                    best as u8
                })
                .collect()
        })
        .collect();
    (g, remap)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::synth::{synth_stack, SynthParams};

    fn max_disp(a: &[Oklab], b: &[Oklab]) -> f64 {
        a.iter().zip(b).map(|(p, q)| dist_sq(*p, *q).sqrt()).fold(0.0, f64::max)
    }
    fn pal() -> Vec<Oklab> {
        synth_stack(&SynthParams::default(), 1, 64).frames[0].palette.clone()
    }

    #[test]
    fn neutral_is_identity() {
        let p = pal();
        assert!(max_disp(&affine_look(&LookCode::neutral(), &p), &p) < 1e-9);
    }

    #[test]
    fn gamut_closure() {
        let p = pal();
        for code in [[1.0, 1.0, 1.0, 1.0], [-1.0, -1.0, -1.0, -1.0], [0.5, -0.7, 0.3, -0.2]] {
            let out = affine_look(&LookCode(code), &p);
            assert!(out.iter().all(|c| (0.0..=1.0).contains(&c[0]) && c[1].abs() <= 0.4 + 1e-9 && c[2].abs() <= 0.4 + 1e-9));
        }
    }

    #[test]
    fn bounded_and_continuous() {
        let p = pal();
        let s = LookCode([0.6, -0.3, 0.8, 0.1]);
        let s2 = LookCode([0.65, -0.3, 0.8, 0.1]);
        assert!(max_disp(&affine_look(&s, &p), &p) <= LOOK_BOUND_MAX + 1e-9);
        let ds = ((s.0[0] - s2.0[0]).powi(2)).sqrt();
        assert!(max_disp(&affine_look(&s, &p), &affine_look(&s2, &p)) <= LOOK_LIPSCHITZ * ds + 1e-9);
    }

    #[test]
    fn collapse_is_valid() {
        let stk = synth_stack(&SynthParams::default(), 8, 64);
        let (g, remap) = global_collapse(&stk, 64, 8);
        assert_eq!(g.len(), 64);
        assert!(remap.iter().all(|f| f.iter().all(|&i| (i as usize) < 64)));
    }
}

File: spec/src/SixFour/Spec/Look.hs
{- |
Module      : SixFour.Spec.Look
Description : The math-verifiable LOOK contract (MATH.md §9).

A per-user "look" model (to be trained on MLX) maps a 64³ GIF's global palette
to a stylised one. The /aesthetics/ are learned and not verifiable — but the
transform's BEHAVIOUR is, and that behaviour is exactly what makes the shipped
control safely user-modulable. This module hardens that contract:

  * 'LookCode' — a bounded control vector in @[-1,1]^d@. It is
    **control-surface-agnostic**: named perceptual knobs, the model's raw
    latent, or a transformer-on-top all map /into/ this code. We freeze the
    code's algebra; the surface above it can iterate.
  * 'LookTransform' — any pure @LookCode -> Palette -> Palette@. A trained net
    is one inhabitant; 'affineLook' is the reference (non-learned) inhabitant.
  * Four laws every conforming transform must satisfy, **independent of
    weights**: neutral-identity (reset works), gamut-closure (no invalid
    colours), boundedness (a bounded knob has bounded effect), and continuity
    (a small knob move is a small look change). A residual, tanh-bounded,
    clamped architecture makes them hold by construction; the laws verify it.

The laws live here (not "SixFour.Spec.Laws") to stay self-contained.
-}
-- COMPARTMENT: MLX-MODEL | tag:MacTag
module SixFour.Spec.Look
  ( LookCode
  , unLookCode
  , mkLookCode
  , neutralLookCode
  , lookDim
  , LookTransform(..)
  , affineLook
  , lookBoundMax
  , lookLipschitz
  , lawLookNeutralIdentity
  , lawLookGamutClosure
  , lawLookBounded
  , lawLookContinuity
  ) where

import qualified Data.Vector as V

import SixFour.Spec.Color   (OKLab(..))
import SixFour.Spec.Palette (Palette(..))

-- | Reference control dimensionality (warmth, contrast, saturation, lift).
-- The analysis put the look manifold at ~5 dims; 4 named axes is the reference.
lookDim :: Int
lookDim = 4

-- | A bounded control vector in @[-1,1]^lookDim@. The neutral code is the
-- origin (the faithful, un-stylised baseline).
newtype LookCode = LookCode
  { unLookCode :: V.Vector Double  -- ^ the raw @[-1,1]^lookDim@ control vector
  }
  deriving (Eq, Show)

-- | Build a code; rejects wrong length or out-of-box values.
mkLookCode :: [Double] -> Maybe LookCode
mkLookCode xs
  | length xs == lookDim && all (\x -> x >= -1 && x <= 1) xs = Just (LookCode (V.fromList xs))
  | otherwise = Nothing

-- | The neutral code: the origin of the box — the faithful, un-stylised baseline.
neutralLookCode :: LookCode
neutralLookCode = LookCode (V.replicate lookDim 0)

-- | A look transform: a pure, palette-size-polymorphic function. A trained
-- model is one inhabitant; 'affineLook' is the reference inhabitant.
newtype LookTransform =
  LookTransform { applyLook :: forall k. LookCode -> Palette k -> Palette k }

-- | Reference (non-learned) inhabitant: a residual, clamped colour-space warp.
-- Neutral code ⇒ identity (each term is 0 / unit-scale at code 0).
affineLook :: LookTransform
affineLook = LookTransform $ \(LookCode v) (Palette ps) ->
  let warmth   = tanh (v V.! 0) * 0.08
      contrast = 1 + 0.5 * tanh (v V.! 1)
      sat      = 1 + 0.5 * tanh (v V.! 2)
      lift     = tanh (v V.! 3) * 0.10
      cl lo hi x = max lo (min hi x)
      warp (OKLab l a b) =
        OKLab (cl 0 1     (0.5 + (l - 0.5) * contrast + lift))
              (cl (-0.4) 0.4 (a * sat + warmth))
              (cl (-0.4) 0.4 (b * sat + warmth))
  in Palette (V.map warp ps)

-- | Generous absolute bound on per-colour displacement over the whole code box.
lookBoundMax :: Double
lookBoundMax = 0.8

-- | Lipschitz constant of the transform w.r.t. the code.
lookLipschitz :: Double
lookLipschitz = 1.5

-- internal: max per-colour OKLab distance between two palettes.
paletteDistMax :: Palette k -> Palette k -> Double
paletteDistMax (Palette a) (Palette b)
  | V.null a  = 0
  | otherwise = V.maximum (V.zipWith d a b)
  where
    d (OKLab l1 a1 b1) (OKLab l2 a2 b2) =
      sqrt ((l1 - l2) ^ (2 :: Int) + (a1 - a2) ^ (2 :: Int) + (b1 - b2) ^ (2 :: Int))

-- | Law (neutral identity): the no-look code returns the palette unchanged
-- (so 'reset' always recovers the faithful baseline).
lawLookNeutralIdentity :: Double -> LookTransform -> Palette k -> Bool
lawLookNeutralIdentity tol lt p = paletteDistMax (applyLook lt neutralLookCode p) p <= tol

-- | Law (gamut closure): the output is always in valid OKLab range.
lawLookGamutClosure :: LookTransform -> LookCode -> Palette k -> Bool
lawLookGamutClosure lt s p =
  let Palette ps = applyLook lt s p
  in V.all (\(OKLab l a b) ->
              l >= -1e-9 && l <= 1 + 1e-9 && abs a <= 0.4 + 1e-9 && abs b <= 0.4 + 1e-9)
           ps

-- | Law (boundedness): a code in the box moves any colour by at most 'lookBoundMax'.
lawLookBounded :: LookTransform -> LookCode -> Palette k -> Bool
lawLookBounded lt s p = paletteDistMax (applyLook lt s p) p <= lookBoundMax + 1e-9

-- | Law (continuity): the transform is Lipschitz in the code — small knob move,
-- small look change. The guarantee that makes a slider feel predictable.
lawLookContinuity :: LookTransform -> LookCode -> LookCode -> Palette k -> Bool
lawLookContinuity lt s s' p =
  let ds = sqrt (V.sum (V.zipWith (\x y -> (x - y) ^ (2 :: Int)) (unLookCode s) (unLookCode s')))
      dp = paletteDistMax (applyLook lt s p) (applyLook lt s' p)
  in dp <= lookLipschitz * ds + 1e-9
